### ISSUE TITLE
[no-unused-expressions] Using throw on same line as if statement triggers runaway EsLint process

### ISSUE THREAD

**Issue by kennethkeim** [typescript-eslint/typescript-eslint#3610]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

I'm using VsCode EsLint plugin, and have been getting runaway "Code Helper (Renderer)" processes for weeks. I finally pinned it down to a specific line that reliably triggers it.

Related to: https://github.com/typescript-eslint/typescript-eslint/issues/1042

**Repro**

I don't have this rule (no-unused-expressions) enabled explicitly, but I'm extending the following. Let me know if you want my full config. I don't have any type-aware rules. Eslint extends prettier, but prettier does not run in the background, it simply disables some eslint rules.
```JSON
extends: [
  "standard-with-typescript",
  "prettier"
],
```

```TS
// this is fine
if (!range) {
   throw new Error("Date range must be set.")
}

// this triggers the runaway process
if (!range) throw new Error("Date range must be set.")
```

**Expected Result**

Nothing, EsLint should continue to run in the background as before.

**Actual Result**

When I type the throw statement on the same line as an if statement (see example code), the "Code Helper (Renderer)" process immediately spikes to 90%+ and won't stop even after killing VsCode. The erorr shown on screen is "Expected an assignment or function call and instead saw an expression.” from eslint@typescript-eslint/no-unused-expressions.

I've looked at the pid and confirmed the process is the EsLint server.

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `^4.28.2` |
| `@typescript-eslint/parser`        | `^4.28.2` |
| `TypeScript`                       | `^4.0.5` |
| `ESLint`                           | `^7.26.0` |
| `node`                             | `14.9` |

**Comment by bradzacher** [comment]
There's likely something else at play as I'm unable to repro.

![image](https://user-images.githubusercontent.com/7462525/124645897-d5bee080-de48-11eb-82ac-ac116f5f7a41.png)

Please create an isolated repro to help with investigation.

**Comment by kennethkeim** [comment]
Well after an hour or so of trying everything else, (testing in multiple of my projects and starting a test repo), I've found the simple way to reproduce it.

There must be a line of code below the if/throw statement
```
function foo(bar: boolean): void {
  if (bar) throw new Error();

  console.log("foo");
}
```

AND

The following rule must be enabled:

```
"padding-line-between-statements": [
      "warn",
      // one of the following rules must be enabled (i assume any "padding-line-between-statements" rule)
      // { blankLine: "always", prev: "multiline-block-like", next: "*" },
      { blankLine: "always", prev: "for", next: "*" },
],
```

So it doesn't look like typescript-eslint is the problem.

**Comment by kennethkeim** [comment]
I should add that the runaway process is triggered once while writing the if/throw line, but doesn't cause any problems afterwards if you leave it in the code after dealing with the runaway process.

**Comment by bradzacher** [comment]
Even with that code, I'm unable to repro it.
The process quickly settles back to 0 when I finish typing:

https://user-images.githubusercontent.com/7462525/124831836-a1265400-df30-11eb-8a1d-0aa9f9a2b4af.mov

**Comment by kennethkeim** [comment]
https://github.com/kennethkeim/ts-eslint-3610

**Comment by kennethkeim** [comment]
To trigger the runaway process, the console.log line has to be in place before you start typing the if/throw line.

**Comment by kennethkeim** [comment]
I've been using the mac os Activity Monitor app so far, just checked the vs code process explorer and it hasn't shown up in there but is definitely using all the CPU in Activity Monitor.

**Comment by JoshuaKGoldberg** [comment]
I can't reproduce this either 😕.

cc @kennethkeim - with only one report from one user and two of us unable to repro, I'm going to close this as non-actionable. If you or anybody else can reproduce this happening, please post back and we can take another look.

**Comment by kennethkeim** [comment]
> To trigger the runaway process, the console.log line has to be in place before you start typing the if/throw line.

Ok that's fine, I just disabled some rules in order to bypass this. Just checking though, did you have the console.log in place as mentioned above?

**Comment by JoshuaKGoldberg** [comment]
Yup, and I cloned your repro and everything.

**Comment by kennethkeim** [comment]
Ok thanks for looking into it
